docs/man/rpm-macros.7.scd

```diff
@@ -0,0 +1,831 @@
+*macros.foo* and *macros['foo']* are equivalent, use what better suits the
+purpose. Like any real Lua table, non-existent items are returned as *nil*,
+and assignment can be used to define or undefine macros.
+
+Example:
+```
+if not macros.yours then
+    macros.my = 'my macro'
+end
+
+local v = { '_libdir', '_bindir', '_xbindir' }
+for _, v in ipairs(v) do
+    if not macros[v] then
+        macros[v] = 'default'
+    end
+end
+```
+
+Parametric macros (including all built-in macros) can be called in a Lua
+native manner via the *macros* table, with either macros.name or macros[name]
+syntax. The argument can be either a single string (*macros.with('thing')*),
+in which case it's expanded and split with the macro-native rules,
+or it can be a table *macros.dostuff({'one', 'two', 'three'})* in which
+case the table contents are used as literal arguments that are not expanded
+in any way.
+
+Lua macros can also also *return* their output, which makes programming
+helper macros look more natural.
+
+Example:
+```
+%sum() %{lua:
+   local v = 0
+   for _, a in ipairs(arg) do
+       v = v + tonumber(a)
+   end
+   return v
+}
+```
+
+# EXAMPLES
+## Example 1. Define a simple macro
+Define macro *mylib* to a path relative to *%{\_libdir}* macro in a spec:
+
+```
+%define mylib %{_libdir}/mine
+```
+
+## Example 2. Define a parametric macro
+Define parametric macro *myhelper* which executes the program
+specified by *%myprog* with it's first argument and always passing
+the option *--some-opt* to it, and additionally the *--xtra* option
+if it received the *-x* option itself:
+
+```
+%define myhelper(x) %{myprog} --some-opt %{?-x:--xtra} %{1}
+```
+
+## Example 3. Define a macro utilizing shell expansion
+Define macro *%today* that expands to the current date in _YYMMDD_
+format by calling the *date*(1) shell utility.
+Note the 2nd *%* needed to escape the arguments to *date*(1):
+
+```
+%define today %(date +%%y%%m%%d)
+```
+
+## Example 4. Define a macro conditionally
+Define macro *mypath* if it wasn't previously defined:
+
+```
+%{!?mypath: %define mypath /some/where}
+```
+
+## Example 5. Conditional expansion
+Expands to *1* if *use_foo* is defined and *0* otherwise:
+
+```
+%{?use_foo:1}%{!?use_foo:0}
+```
+
+## Example 6. Expressions
+Calculate 5 \* 1024:
+
+```
+%[5 * 1024]
+```
+
+Expand to literal *true* or *false* depending on a condition:
+
+```
+%[1 < 2 ? "true" : "false"]
+```
+
+Compare versions, expanding to *1* or *0* on true/false:
+
+```
+%[ v"3.1.0-1" < v"1.0~alpha-2" ? 1 : 0]
+```
+
+Expands to *1* if *%aa* expands to *5*, otherwise expands to *2*:
+
+```
+%[ "%{aa}" == "5" ? 1 : 2]
+```
+
+## Example 7. Define a parametric Lua macro
+Define *%range* Lua macro that takes two integers as arguments and
+expands to the catenation of the numbers in the range:
+
+```
+%define range() %{lua:
+    for i=math.tointeger(arg[1]),math.tointeger(arg[2]) do
+        print(i)
+    end
+}
+```
+
```

Comment:
```
> %range 4 8
45678
```
May be this isn't that great of an example. may be add `.." "` to the print even if this appends an unwanted (?) whitespace at the end.

Reply:
Dropped this example from the main macro manual, it belongs to the to-be-added rpm-lua(7) instead.
We can try coming up with a better example there...